quinn-proto/src/connection/streams/recv.rs

```diff
@@ -257,71 +198,179 @@ impl Recv {
+            id,
+            recv: Some(rs),
+            streams,
+            pending,
+            ordered,
+            read: 0,
+            fused: None,
+        })
     }
-}
 
-pub(crate) struct ReadChunks {
-    pub bufs: usize,
-    pub read: usize,
+    /// Read the next chunk from the stream
+    ///
+    /// Return the next chunk of data from the stream, if any are available. `max_length` may be
+    /// used to restrict the maximum size of the returned `Bytes` (but `usize::MAX` is supported,
+    /// too).
+    pub fn next(&mut self, max_length: usize) -> ReadResult {
+        if let Some(status) = &self.fused {
+            return ReadResult::from_status(*status);
+        }
+
+        let rs = self.recv.as_mut().unwrap();
+        let chunk = rs.assembler.read(max_length, self.ordered);
+        let status = match &chunk {
+            Some(chunk) => {
+                self.read += chunk.bytes.len() as u64;
+                ReadStatus::Readable
+            }
+            None => match rs.state {
+                RecvState::Recv { size } => {
+                    if size == Some(rs.end) && rs.assembler.bytes_read() == rs.end {
+                        self.streams.stream_freed(self.id, StreamHalf::Recv);
+                        self.fused = Some(ReadStatus::Finished);
+                        ReadStatus::Finished
+                    } else {
+                        self.streams.recv.insert(self.id, self.recv.take().unwrap());
+                        self.fused = Some(ReadStatus::Blocked);
+                        ReadStatus::Blocked
+                    }
+                }
+                RecvState::ResetRecvd { error_code, .. } => {
+                    rs.state = RecvState::Closed;
+                    self.streams.stream_freed(self.id, StreamHalf::Recv);
+                    self.fused = Some(ReadStatus::Reset(error_code));
+                    ReadStatus::Reset(error_code)
+                }
+                RecvState::Closed => unreachable!(),
+            },
+        };
+
+        let mut should_transmit = ShouldTransmit(false);
+        if status != ReadStatus::Readable {
+            should_transmit = self.finalize();
+        }
+
+        ReadResult {
+            chunk,
+            status,
+            should_transmit,
+        }
+    }
+
+    /// Finish reading, letting the caller know if the connection has something to transmit
+    pub fn finalize(&mut self) -> ShouldTransmit {
+        let mut rs = match self.recv.take() {
+            Some(rs) => rs,
+            None => return ShouldTransmit(false),
+        };
+
+        let (_, max_stream_data) = rs.max_stream_data(self.streams.stream_receive_window);
+        let max_data = self.streams.add_read_credits(self.read);
+        let max_dirty = self.streams.take_max_streams_dirty(self.id.dir());
+        self.pending
+            .post_read(self.id, max_data, max_stream_data, max_dirty);
+        max_data | max_stream_data | ShouldTransmit(max_dirty)
```

Comment:
Weren't those 2 separate code path before? One one whether max data or max stream data frames could be converted, and one for max streams. I guess if it's only about signalling that something needs to be transmitted then unifying them is great. But if something makes a distinction what to transmit based on the return value it might not work.

Reply:
The `pending.post_read()` call should ensure that both are sent; returning the `ShouldTransmit` to the caller is just about helping the caller decide whether calling `poll_transmit()` is necessary.